File: concatenate.php
<?php
/**
 * concatenate.php
 * Concatenates the Netflix Prize traning_set files into one single file. 
 * Input arguments: [1] : Directory of traning_set files
 *                  [2] : Path to output file. If file does not exist, 
 *                        the script will create a new one.
 * The script formats the concatenated in following:
 * [item id],[user id],[rating]
 * [item id],[user id],[rating]
 * ...
 * Date is an optional variable, dependent on use (i.e, Temporal dynamics).
 */

$time_start = microtime(true);

$dirWithMovies = $_SERVER["argv"][1];
$outFile = $_SERVER["argv"][2];

is_dir($dirWithMovies) 
  or die($dirWithMovies . " is not a directory.\n");

$dh = opendir($dirWithMovies)
  or die("Error opening directory: " . $dirWithMovies);

$ofile = fopen($outFile, "w");

while (($file = readdir($dh)) != FALSE) {
  $file = $dirWithMovies . "/" . $file;
  $fc = file($file);

  $itemID = array_shift($fc);
  $itemID= rtrim($itemID);
  $itemID = rtrim($itemID,":");
  
  foreach ($fc as $line) {
    $pieces = explode(',', $line);
    $userID = $pieces[0];
    $rating = $pieces[1];
    //$date = $pieces[2];

    $outLine =  $itemID . ','  . $userID . ',' . $rating .  "\n";
    fwrite($ofile, $outLine) 
      or die("Error writing to file " . $outFile . "\n");
  }

}

closedir($dh);

$time_end = microtime(true);
$time = $time_end - $time_start;
echo "Runtime: ".round($time,2)." seconds\n";
?>

File: concatenate.pl
#!/usr/bin/perl -w
use warnings;
use strict;

# concatenate.pl
# Concatenates the Netflix Prize traning_set files into one single file.
#

# Quit unless we have the correct number of command-line args
if (($#ARGV+1) != 2) {
    print "Please enter right number of arguments: 'perl scriptname.pl [Directory of Raw Data] [Output directory]'\n";
    exit;
}

# The directory to scan containing the raw files 
my $dirRaw = $ARGV[0];
my $dirOutput = $ARGV[1];

opendir DIR, $dirRaw 
	or die("Could not open $dirRaw");

my $outString;
my $outFile;
while(my $fname = readdir DIR) {
        my $fname = "$dirRaw/$fname";
        open FILE, $fname 
        	or die("Could not open $fname");
        (my $mid = <FILE>) =~ s/:.*//s;
        while(<FILE>) {
                chomp;
                $outString = qq($mid,).qq($_)."\n";
                open $outFile, ">>", "$dirOutput/concat_raw_data.txt"
                	or die ("Error opening to new file.");
                print {$outFile} $outString;
        }
        close $outFile;
        close FILE;
}
closedir DIR;
print "Done.";
exit;